org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/JavaRuntime.java

```diff
@@ -1232,7 +1233,9 @@ public static IRuntimeClasspathEntry[] resolveRuntimeClasspathEntry(IRuntimeClas
 						return new IRuntimeClasspathEntry[0];
 					}
 					IClasspathAttribute[] attributes = entry.getClasspathEntry().getExtraAttributes();
-					IRuntimeClasspathEntry[] entries = resolveOutputLocations(project, entry.getClasspathProperty(), attributes, excludeTestCode);
+					boolean withoutTestCode = entry.getClasspathEntry().isWithoutTestCode();
+					IRuntimeClasspathEntry[] entries = resolveOutputLocations(project, entry.getClasspathProperty(), attributes, excludeTestCode
+							|| withoutTestCode);
```

Comment:
If value is "No" this returns true? 
As I understood, for the value of "No" the dependency should be excluded.

Reply:
Test code should be excluded in `resolveOutputLocations` when one of these conditions is met:

1. It is explicitly requested in the *launch configuration* (This is the existing condition.):
```
boolean excludeTestCode = configuration.getAttribute(IJavaLaunchConfigurationConstants.ATTR_EXCLUDE_TEST_CODE, false);
```

2. It is stated in the classpath entry of the *project dependency* (The is the new condition introduced with this PR):
```
boolean withoutTestCode = entry.getClasspathEntry().isWithoutTestCode();
```

* When the launch configuration request exclusion of test code (case 1), test code will be excluded as before.
* When the launch configuration does not request exclusion but the current classpath entry is a project dependency and this entry is configured to not expose test code `entry.getClasspathEntry().isWithoutTestCode() `evaluates to **`true`** (case 2), so test code should also be excluded.

Currently, the `without_test_code` attribute of the classpath entry is not evaluated when resolving the runtime classpath. When resolving the classpath in the compiler, this attribute is already examined:

https://github.com/eclipse-jdt/eclipse.jdt.core/blob/a8bfcc50993854c9584914d21b9a77427c6ebb2d/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaProject.java#L602

```
boolean nestedWithoutTestCode = excludeTestCode || entry.isWithoutTestCode();
```